azkaban-common/src/main/java/azkaban/flow/Flow.java

```diff
@@ -171,29 +171,32 @@ public void initialize() {
         }
       }
 
-      for (final Node node : this.startNodes) {
-        node.setLevel(0);
-        this.numLevels = 0;
-        recursiveSetLevels(node);
-      }
+      setLevelsAndEdgeNodes(new HashSet<>(startNodes), 0);
     }
   }
 
-  private void recursiveSetLevels(final Node node) {
-    final Set<Edge> edges = this.outEdges.get(node.getId());
-    if (edges != null) {
-      for (final Edge edge : edges) {
-        final Node nextNode = this.nodes.get(edge.getTargetId());
-        edge.setSource(node);
-        edge.setTarget(nextNode);
+  private void setLevelsAndEdgeNodes(final Set<Node> levelNodes, int level) {
+    final Set<Node> nextLevelNodes = new HashSet<>();
+
+    for (Node node : levelNodes) {
```

Comment:
Node node -> final Node node?
nodes -> this.nodes?
Again, these can be auto checked by "save action" plugin. :)